### ISSUE TITLE
[Question] Importance of backward pass in root finder

### ISSUE THREAD

**Issue by rydeveraumn** [google-research/neural-structural-optimization#2]
Hey @shoyer! I do see this repository has been dormant for quite some time. My name is Ryan and I am interested in implementing this in `Pytorch`. It seems that I have just about everything in place. However, it seems my `x` values become very very large after a first backward pass so leads me to believe that the calculation of the adjoint derivative is actually fairly important.

This is however where my understanding breaks down. If you have a moment to help me understand how that works under the hood I would very much appreciate it or if you have any tips on a `pytorch` implementation (don't know if you have experience with it after all these years) I would appreciate that as well.

Thank you for your time in advance and this is some really neat work!

**Comment by shoyer** [comment]
Yes, you need to take care to calculate custom gradients. I imagine you should be able to directly adapt the [custom gradient rules](https://github.com/google-research/neural-structural-optimization/blob/master/neural_structural_optimization/autograd_lib.py) that we wrote for autograd into PyTorch.

**Comment by rydeveraumn** [comment]
Hey @shoyer wow thanks for the quick response! I do have one more question from your perspective do you think it would be possible to place general constraints on `x` without actually having to define custom gradients. I have software where the goal is to solve general constraint problems.

So the idea would be to add a penalty to x to be within [0, 1] and also have another constraint where the mean of the physical density - V = 0. 

Example:
https://github.com/sun-umn/PyGRANSO/blob/main/examples/A1_rosenbrock.ipynb

Again thanks for your assistance!

**Comment by shoyer** [comment]
You should look for a PyTorch equivalent to JAXopt for solving numerical equations with constraints, i.e., the equivalent of `scipy.optimize`. Maybe xitorch? https://github.com/xitorch/xitorch

**Comment by rydeveraumn** [comment]
Thanks @shoyer! I appreciate the assistance.
